File: src/formatter.ts
import type { LintReport } from "./types.js";

export interface FormatOptions {
  color?: boolean;
  verbose?: boolean;
  helpUrl?: string;
}

const color = {
  red: (s: string) => `\x1b[31m${s}\x1b[39m`,
  yellow: (s: string) => `\x1b[33m${s}\x1b[39m`,
  green: (s: string) => `\x1b[32m${s}\x1b[39m`,
  bold: (s: string) => `\x1b[1m${s}\x1b[22m`,
  dim: (s: string) => `\x1b[2m${s}\x1b[22m`,
};

const identity = (s: string) => s;

export function formatReport(report: LintReport, options: FormatOptions = {}): string {
  const useColor = options.color !== false;
  const red = useColor ? color.red : identity;
  const yellow = useColor ? color.yellow : identity;
  const green = useColor ? color.green : identity;
  const bold = useColor ? color.bold : identity;
  const dim = useColor ? color.dim : identity;

  const lines: string[] = [];

  for (const outcome of report.results) {
    if (outcome.valid && !options.verbose) continue;

    lines.push(`${bold("⧗")}   input: ${outcome.input.split("\n")[0]}`);

    for (const err of outcome.errors) {
      lines.push(`  ${red("✖")}   ${err.message} ${dim(`[${err.name}]`)}`);
    }
    for (const warn of outcome.warnings) {
      lines.push(`  ${yellow("⚠")}   ${warn.message} ${dim(`[${warn.name}]`)}`);
    }

    lines.push("");
  }

  if (report.errorCount > 0 || report.warningCount > 0) {
    const summary = `${report.errorCount} error(s), ${report.warningCount} warning(s)`;
    const colorizedSummary = report.errorCount > 0 ? red(summary) : yellow(summary);
    lines.push(`${bold("✖")}   found ${colorizedSummary}`);

    if (options.helpUrl) {
      lines.push(`${dim("ℹ")}   Get help: ${options.helpUrl}`);
    }
  } else if (options.verbose) {
    lines.push(`${green("✔")}   0 problems, 0 warnings`);
  }

  return lines.join("\n");
}

export default formatReport;

File: src/formatter.rs
use crate::linter::LintOutcome;

pub struct FormatOptions {
    pub color: bool,
    pub verbose: bool,
    pub help_url: Option<String>,
}

impl Default for FormatOptions {
    fn default() -> Self {
        Self {
            color: true,
            verbose: false,
            help_url: None,
        }
    }
}

pub fn format_report(report: &LintOutcome, options: &FormatOptions) -> String {
    let red = |s: &str| {
        if options.color {
            format!("\x1b[31m{}\x1b[39m", s)
        } else {
            s.to_string()
        }
    };
    let yellow = |s: &str| {
        if options.color {
            format!("\x1b[33m{}\x1b[39m", s)
        } else {
            s.to_string()
        }
    };
    let green = |s: &str| {
        if options.color {
            format!("\x1b[32m{}\x1b[39m", s)
        } else {
            s.to_string()
        }
    };
    let bold = |s: &str| {
        if options.color {
            format!("\x1b[1m{}\x1b[22m", s)
        } else {
            s.to_string()
        }
    };
    let dim = |s: &str| {
        if options.color {
            format!("\x1b[2m{}\x1b[22m", s)
        } else {
            s.to_string()
        }
    };

    let mut lines = Vec::new();

    if !report.valid || options.verbose {
        let first_line = report.input.lines().next().unwrap_or("");
        lines.push(format!("{}   input: {}", bold("⧗"), first_line));

        for err in &report.errors {
            lines.push(format!(
                "  {}   {} {}",
                red("✖"),
                err.message,
                dim(&format!("[{}]", err.name))
            ));
        }
        for warn in &report.warnings {
            lines.push(format!(
                "  {}   {} {}",
                yellow("⚠"),
                warn.message,
                dim(&format!("[{}]", warn.name))
            ));
        }
        lines.push(String::new());
    }

    if !report.errors.is_empty() || !report.warnings.is_empty() {
        let summary = format!(
            "{} error(s), {} warning(s)",
            report.errors.len(),
            report.warnings.len()
        );
        let colorized = if !report.errors.is_empty() {
            red(&summary)
        } else {
            yellow(&summary)
        };
        lines.push(format!("{}   found {}", bold("✖"), colorized));

        if let Some(help) = &options.help_url {
            lines.push(format!("{}   Get help: {}", dim("ℹ"), help));
        }
    } else if options.verbose {
        lines.push(format!("{}   0 problems, 0 warnings", green("✔")));
    }

    lines.join("\n")
}
